### DISCUSSION TITLE
Heroku + S3 = img's with expiry

### DISCUSSION THREAD

**Discussion by ddvorkin14** [solidusio/solidus#4669] [Troubleshooting]
There is no clear documentation around how to properly setup the s3 bucket and the solidus instance on heroku so img's dont expire.... Whats strange is the img's dont expire on the solidus backend panel, its only on the customer facing front end views

**Solidus Version**:
3.1.7

**To Reproduce**
deploy to heroku with s3 config, upload an img, visit the customer facing ui, wait 5 min and the img will expire and will not display

**Current behavior**
the img link is expiring on the frontend and its a bit confusing why

**Expected behavior**
Im expecting for the url to the img to refresh or reload or something

**Screenshots**

![Screen Shot 2022-10-04 at 10 15 01 AM](https://user-images.githubusercontent.com/102325794/193843050-87c8c642-6054-44d2-bda3-def0258cf1c3.png)

**Desktop (please complete the following information):**
 - OS: HEROKU
 - Browser Chrome

**Comment by jarednorman** [comment]
Yeah, this is one of the most common issues people run into that gets reported in the Solidus Slack #support channel. The backend isn't an issue because there's no caching.

The frontend runs into the issue that much of the product show and listing pages are cached, so when the expiring URLs are generated, they get reused beyond the lifespan of the generated image URLs.

**Comment by ddvorkin14** [comment]
So what is the recommendation here because im running out of ideas haha @jarednorman

**Comment by jarednorman** [comment]
If you set the bucket policy to make all objects in the bucket public, that'll solve it.

**Reply by elia** [reply]
@ddvorkin14 For more info you can refer to this section of the active storage guide https://edgeguides.rubyonrails.org/active_storage_overview.html#public-access

If you have already a bucket set up be sure to properly migrate it to public:

> Make sure your buckets are properly configured for public access. See docs on how to enable public read permissions for [Amazon S3](https://docs.aws.amazon.com/AmazonS3/latest/user-guide/block-public-access-bucket.html), [Google Cloud Storage](https://cloud.google.com/storage/docs/access-control/making-data-public#buckets), and [Microsoft Azure](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-manage-access-to-resources#set-container-public-access-level-in-the-azure-portal) storage services. Amazon S3 additionally requires that you have the s3:PutObjectAcl permission.

**Comment by erwin** [comment]
> set the bucket policy to make all objects in the bucket public

Are there any alternatives to setting `config.active_storage.service` to a S3 bucket that you've configured with `public: true` to get cached pages to work?

For example, I do not want images and other media attached to my applications ActionText RichText fields to also be public, and `config.active_storage.service` seems to be the only way to handle those...

1) Could I create a monkeypatch in `config/initializers` that would update the `Spree::Image::ActiveSupportAttachment`'s `has_attachment` to include something like `service: :s3_public`?

2) Could I extend the `expires_in` time of the S3 requests so they match the cache expiration time?

3) Can I disable some parts of the fragment caching or time it out more quickly so the private S3 links can continue to work?


So far the only sure fire thing I've thought of is to split my app into two parts and serve them up seperately, but that's very messy, a whole lot of work, and I would absolutely like to avoid it possible...
